stream_alert_cli/test.py

```diff
@@ -281,24 +284,75 @@ def test_rule(rule_name, test_record, formatted_record):
                 record for the service to be tested
 
         Returns:
-            [bool] boolean indicating if this rule passed
+            [list] alerts that hit for this rule
+            [integer] count of expected alerts for this rule
+            [bool] boolean where False indicates errors occurred during processing
         """
         event = {'Records': [formatted_record]}
 
         expected_alert_count = test_record.get('trigger_count')
         if not expected_alert_count:
-            expected_alert_count = (0, 1)[test_record['trigger']]
+            expected_alert_count = 1 if test_record['trigger'] else 0
+
+        # Run the rule processor. Passing mocked context object with fake
+        # values and False for suppressing sending of alerts
+        processor = StreamAlert(self.context, False)
+        all_valid_logs = processor.run(event)
 
-        # Run the rule processor. Passing 'None' for context
-        # will load a mocked object later
-        alerts = StreamAlert(None, True).run(event)
+        if not all_valid_logs:
+            payload = StreamPayload(raw_record=formatted_record)
+            classifier = StreamClassifier(config=load_config())
+            classifier.map_source(payload)
+            logs = classifier._log_metadata()
+            self.analyze_log_delta(logs, rule_name, test_record)
+
+        alerts = processor.get_alerts()
 
         # we only want alerts for the specific rule being tested
         alerts = [alert for alert in alerts
                   if alert['rule_name'] == rule_name]
 
-        return alerts, expected_alert_count
+        return alerts, expected_alert_count, all_valid_logs
+
+    def analyze_log_delta(self, logs, rule_name, test_record):
```

Comment:
Confusing docblock. The first sentence makes it sound like this function should be used to describe why a test is failing. Aka you already know it is. But, as you read on, you realize the function's job is to determine *if* a test will fail. I think? Please fix this and choose a better function name. Also, the docblock mentions records, but the variables mention logs. Consistency please.